### DISCUSSION TITLE
播放YouTube视频很卡，希望适配好aria2

### DISCUSSION THREAD

**Discussion by machine-doll** [dyphire/mpv-config#6] [Q&A] [answered]
试了一下自己修改youtube-download.lua
第85行，改成了` youtube_dl_exe = "./youtube-dl.exe --external-downloader ./aria2c.exe --external-downloader-args \"-x 16 -s 16\"",`
但是播放视频还是很卡，不确定有没有调用aria2，我已经把aria2放到mpv.net的文件夹里了
之前试过给youtube-dl提速，这样子做是可以的。尝试过多种参数组合，都不能够提升播放视频的流畅度
另外，第35行，我改成了`video_format = "av01,av1,vp9",`还是不能够优先使用av1编码，依旧在请求h264的格式

**Comment by dyphire** [comment, marked as answer]
你搞错脚本功能了，youtube-download.lua是用来下载ytdl视频到本地的脚本，并不是加载时使用的。
quality-menu.lua是ytdl加载后用来选择音视频质量的菜单脚本。

而mpv用来加载http视频的功能实际是使用内置的ytdl-hook.lua脚本实现的，我禁用了内置脚本，并用增强脚本ytdl_hook_plus.lua替代。
如果你想进行进一步修改并增强ytdl功能应该调整ytdl_hook_plus.lua脚本。

另外YouTube视频加载慢很大程度上是因为你使用了youtube-dl.exe，youtube-dl开发已经很不活跃了，release更不积极，一直不更新。建议换用维护更活跃的[yt-dlp](https://github.com/yt-dlp/yt-dlp)分支，加载方面会有明显改善。ytdl-hook.lua和ytdl_hook_plus.lua其实默认优先查找yt-dlp，不应选择使用youtube-dl。yt-dlp的每日构建：https://github.com/ytdl-patched/yt-dlp/releases

**Comment by machine-doll** [comment]
我试着修改了一下ytdl_hook_plus.lua，现在av1和vp9的权重比h264高了，也把youtube-dl.exe换成了yt-dlp，但是默认似乎依旧是单线程加载。播放视频的时候依旧有卡顿的现象  
有办法让yt-dlp用外置的aria2c作为下载器并且启动多线程下载吗？我用浏览器播放YouTube视频的时候，YouTube会使用多个线程来获取视频，这种情况下完全不会卡顿。我猜测我的网络单线程下传输速度就是这么低，必须得用多个线程来下载才可以保证流畅播放视频  
另外，我没有找到配置代理服务器的部分，不知道有没有办法添加一下判断是否要用代理服务器以及让yt-dlp之类的程序走代理的功能。我似乎没有找到很好的办法来让程序在启动的时候自己判断。目前是用的netch的tun模式，更常见的http代理我没有什么头绪。youtube-dl单独是可以配置代理的，但是那样就写死了很多东西。有办法传递额外的参数给yt-dlp吗？

**Reply by zhongfly** [reply]
以我的理解是这样的：mpv使用ytdl-hook.lua这样的脚本来调用外部的yt-dlp获取视频文件的地址，然后使用mpv内的ffmpeg依赖来把视频文件**边下边播**（注意，是一边下载一边播放的，或者说是先下载一小部分然后播放、同时缓存下一部分）。而aria2c只是一个下载器，不能实现边下边播。所以ytdl-hook.lua只是一个提供视频文件地址的角色，不能决定视频是如何加载的。这只是我粗浅的理解
想要多线程，可能需要从mpv内部的代码想办法？

**Comment by dyphire** [comment]
> 有办法让yt-dlp用外置的aria2c作为下载器并且启动多线程下载吗？

因为我基本不怎么用mpv/mpv.net观看在线视频，所以对ytdl相关功能研究不深，这方面只能说爱莫能助了。

> 另外，我没有找到配置代理服务器的部分，不知道有没有办法添加一下判断是否要用代理服务器以及让yt-dlp之类的程序走代理的功能
有办法传递额外的参数给yt-dlp吗？

传递给yt-dlp/youtube-dl的代理等额外参数其实是在mpv.conf里进行追加的，参考：
https://github.com/dyphire/mpv-config/blob/a79506ddb8c041989d34dace57933c04926f2711/mpv.conf#L428-L437

使用`ytdl-raw-options-append=proxy=`和`ytdl-raw-options-append=cookies=`可以传递http代理和cookies给yt-dlp/youtube-dl。因为ytdl_hook_plus.lua直接继承了ytdl-hook.lua的代码和功能，所以这些参数也同样有效，这里的注释有点小问题。
PS：其实我也是直接走的tun模式
